src/inbox/mod_inbox_muc.erl

```diff
@@ -4,68 +4,104 @@
+    ejabberd_hooks:delete(update_inbox_for_muc, Host, ?MODULE, update_inbox_for_muc, 90),
     ok.
 
-update_inbox(Acc, Room, {From, FromRoomJid}, AffsDict, Packet) ->
-    Affs = dict:to_list(AffsDict),
-    Users = affs_to_allowed_users(Affs),
-    FromBare = jid:to_bare(From),
-    lists:foreach(fun(User) -> update_inbox(Room, 
-                                            {FromBare, FromRoomJid}, 
-                                            User, 
-                                            Packet) end,
-                  Users),
-    Acc.
 
-affs_to_allowed_users(Users) ->
-    AllowedUsers = lists:filter(fun({_User, outcast}) -> false;
-                    (_) -> true end, Users),
-    lists:map(fun({{User, Domain, Res}, _Aff}) -> jid:make(User, Domain, Res) end,
-              AllowedUsers).
+-spec update_inbox_for_muc(Acc) -> Acc when
+      Acc :: map().
+update_inbox_for_muc(
+    #{room_jid := Room,
+      from_jid := From,
+      from_room_jid := FromRoomJid,
+      packet := Packet,
+      affiliations_map := AffsMap} = Acc) ->
+    F = fun(AffLJID, Affiliation) ->
+            case is_allowed_affiliation(Affiliation) of
+                true ->
+                    To = jid:to_bare(jid:make(AffLJID)),
+                    %% Guess direction based on user JIDs
+                    Direction = direction(From, To),
+                    Host = To#jid.lserver,
+                    Packet2 = jlib:replace_from_to(FromRoomJid, To, Packet),
+                    update_inbox_for_user(Direction, Host, Room, To, Packet2);
+                false ->
+                    ok
+            end
+        end,
+    mongoose_maps:maps_foreach(F, AffsMap),
+    Acc.
 
+-spec is_allowed_affiliation(mod_muc:affiliation()) -> boolean().
+is_allowed_affiliation(outcast) -> false;
+is_allowed_affiliation(_)       -> true.
 
-update_inbox(Room, {From, FromRoomJid}, To, Packet) ->
-    Host = To#jid.server,
-    case direction(From, To) of
-        outgoing ->
-            NewPacket = jlib:replace_from_to(FromRoomJid, To, Packet),
-            handle_outgoing_message(Host, From, Room, NewPacket);
-        incoming ->
-            NewPacket = jlib:replace_from_to(FromRoomJid, To, Packet),
-            handle_incoming_message(Host, Room, To, NewPacket)
+-spec update_inbox_for_user(Direction, Host, Room, To, Packet) -> term() when
+      Direction :: incoming | outgoing,
+      Host :: receiver_host(),
+      Room :: room_bare_jid(),
+      To :: receiver_bare_user_jid(),
+      Packet :: packet().
+update_inbox_for_user(Direction, Host, Room, To, Packet) ->
+    %% Check that Host is a local served domain (not s2s)
+    case {is_local_host(Host), Direction} of
+        {true, outgoing} ->
+            handle_outgoing_message(Host, Room, To, Packet);
+        {true, incoming} ->
+            handle_incoming_message(Host, Room, To, Packet);
+        _ ->
+            ok
     end.
 
+%% From and To are user JIDs (not room JIDs)
```

Comment:
As I've said, I would prefer specs to comments here, but this is a story for a different PR probably